pkg/virt-handler/node-labeller/node_labeller.go

```diff
@@ -0,0 +1,282 @@
+			if !ok {
+				tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+				if !ok {
+					log.Log.Reason(fmt.Errorf("couldn't get object from tombstone %+v", obj)).Error("Failed to process delete notification")
+					return
+				}
+				_, ok = tombstone.Obj.(*v1.ConfigMap)
+				if !ok {
+					log.Log.Reason(fmt.Errorf("tombstone contained object that is not a config map %#v", obj)).Error("Failed to process delete notification")
+					return
+				}
+			}
+
+			n.Queue.Add(n.host)
+		},
+		UpdateFunc: func(_, _ interface{}) {
+			n.Queue.Add(n.host)
+		},
+	})
+
+	go n.kubevirtInformer.Run(stop)
+
+	n.nodeInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		UpdateFunc: func(_, _ interface{}) {
+			n.Queue.Add(n.host)
+		},
+	})
+
+	go n.nodeInformer.Run(stop)
+	cache.WaitForCacheSync(stop, n.nodeInformer.HasSynced, n.kubevirtInformer.HasSynced)
+
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(n.runWorker, time.Second, stop)
+	}
+	<-stop
+}
+func (n *NodeLabeller) runWorker() {
+	for n.Execute() {
+	}
+}
+
+func (n *NodeLabeller) Execute() bool {
+	key, quit := n.Queue.Get()
+	if quit {
+		return false
+	}
+	defer n.Queue.Done(key)
+
+	if !n.kvmController.NodeHasDevice(kvmPath) {
+		n.logger.Errorf("Node-labeller cannot work without KVM device.")
+		return true
+	}
+
+	err := n.run()
+
+	if err != nil {
+		n.logger.Errorf("node-labeller sync error encountered: %v", err)
+		n.Queue.AddRateLimited(key)
+	} else {
+		n.Queue.Forget(key)
+	}
+	return true
+}
+
+func (n *NodeLabeller) run() error {
+	cpuFeatures := make(map[string]bool)
+	cpuModels := make([]string, 0)
+
+	//parse all informations
+	cpuModels, cpuFeatures = n.getCPUInfo()
+
+	//load node
+	nodeObj, exists, err := n.nodeInformer.GetStore().GetByKey(n.host)
+	if err != nil || !exists {
+		n.logger.Errorf("node-labeller cannot get node %s", n.host)
+		return err
+	}
+	var (
+		ok           bool
+		originalNode *v1.Node
+	)
+	if originalNode, ok = nodeObj.(*v1.Node); !ok {
+		n.logger.Errorf("node-labeller cannot convert node " + n.host)
+		return fmt.Errorf("Could not convert node " + n.host)
+	}
+
+	node := originalNode.DeepCopy()
+	//prepare new labels
+	newLabels := n.prepareLabels(cpuModels, n.hypervFeatures.items, cpuFeatures)
+	//remove old labeller labels
+	n.removeLabellerLabels(node)
+	//add new labels
+	n.addLabellerLabels(node, newLabels)
+
+	//patch node only if there is change in labels
+	err = n.patchNode(originalNode, node)
+	return err
+}
+
+func (n *NodeLabeller) patchNode(originalNode, node *v1.Node) error {
+	originalLabelsBytes, err := json.Marshal(originalNode.Labels)
+	if err != nil {
+		return err
+	}
+
+	originalAnnotationsBytes, err := json.Marshal(originalNode.Annotations)
+	if err != nil {
+		return err
+	}
+
+	labelsBytes, err := json.Marshal(node.Labels)
+	if err != nil {
+		return err
+	}
+
+	annotationsBytes, err := json.Marshal(node.Annotations)
+	if err != nil {
+		return err
+	}
+
+	if !reflect.DeepEqual(labelsBytes, originalLabelsBytes) {
```

Comment:
Here it may be better to compare the maps, instead of bytes of json encoded maps, because the json encoder could shuffle the order of elements in the map. It may be improbable, but let's not depend on it.